File: src/main/java/me/austin/queer/module/setting/Setting.java
package me.austin.queer.module.setting;

import me.austin.queer.TransRights;
import me.austin.queer.event.client.SettingChangeEvent;
import me.austin.queer.module.IModule;
import me.austin.queer.module.Module;
import me.austin.queer.module.Modules;

public abstract class Setting<T> extends Module {
	public T currentValue;
	public IModule parent;

	public Setting(String name, String description, T defaultValue, IModule parent) {
		this(name, description, defaultValue, parent.getClass());
	}

	public Setting(String name, String description, T defaultValue, Class<? extends IModule> parent) {
		super(name, description);
		this.currentValue = defaultValue;
		this.parent = Modules.getModuleByClass(parent);
		Settings.add(this);
	}
	
	public T get() {
		return this.currentValue;
	}

	public void set(T value) {
		this.currentValue = value;
		TransRights.EVENTBUS.post(new SettingChangeEvent(this));
	}
}

File: src/main/kotlin/trans/rights/client/modules/setting/Setting.kt
package trans.rights.client.modules.setting

import trans.rights.client.modules.Module

abstract class Setting<T : Any>(name: String, default: T, var value: T = default) : Module(name), Comparable<Setting<*>> {
    fun set(other: T) {
        this.value = other
    }

    override fun compareTo(other: Setting<*>): Int {
        return this.name.compareTo(other.name)
    }
}
